lib: ZMQ TODO's

bingen opened the pull request:
Add write callback.
Add error callback.
Add frrzmq_check_events() function to check for edge triggered things
that may have happened after a zmq_send() call or so.
Update ZMQ tests.

bingen commented:
@eqvinox I left indent commit aside because I saw it's changing a lot of things I didn't touch, so maybe it's due to the lack of idempotency of the script. I can remove it if you want (or try to run it several times until it converges back, do you remember the number?)

eqvinox commented:
So... I'm not sure it's useful to separate read and write handlers. ZeroMQ forces the user to obey to its state machine expectations, so the two directions can't really be used independently. Also, error state is really common to the two...

I'm not sure how to make this work best in terms of API, but I'd argue we should only have one `frrzmq_cb` for a ZeroMQ socket. Maybe it always has an error callback, but read / write is controlled separately?

bingen commented:
I don't think I get it. What do you mean by "indenpendently" I think you can use them independently. Actually, the code merge in right now only has read callback, and use it writing too in a separate module, so it's working independently.
The error callback could do different things if it comes from a read try or from a write try. 
The cancelled flag should be duplicated as well, because you might want to cancel only read or only write.
Threads would be different too, so duplicated too.
Write doesn't use the partial callback.
I thought initially of doing it all together, but I discarded it for these reasons.

donaldsharp commented:
@equinox any updates here?

donaldsharp commented on lib/frr_zmq.c:
```diff
@@ -79,13 +82,21 @@ static int frrzmq_read_msg(struct thread *t)
 
 		if (ret < 0)
 			goto out_err;
+		if (polli.revents & ZMQ_POLLERR)
```
http://api.zeromq.org/4-2:zmq-poll

This documentation specifically states that ZMQ_POLLERR -> 'For ØMQ sockets this flag has no effect if set in events, and shall never be returned in revents by zmq_poll().'

Are you actually seeing this returned?

donaldsharp commented:
In general I agree with David.  This code adds handling of read write such that the calling functions need to know more about the underlying code implementation that I think that they probably should care about.  Having said that if no-one else is going to use this except for the pseudo-wire code and the length of this delay we should carefully consider just taking this code

bingen commented:
I don't think that's what David was saying (he was talking about using read and write handlers independently, which I already replied) and I don't agree with what you are saying either.
If you look at the exposed API, in lib/frr_zmq.h, it's practically the same, and it's actually even more complex in this new version. Besides the free/cancellation management becomes more complex.
Anyway, I have created a new PR addressing your suggestions: #1478
